### DISCUSSION TITLE
Ractive template build watcher

### DISCUSSION THREAD

**Discussion by evs-chris** [ractivejs/ractive#3405] [General]
Here's the process that I use to build my views incrementally in my larger projects:

1. I have a script that initializes the output structure of my client code.
2. The script then starts a custom node script (see below) that watches my `views` directory and spits out ts modules in a gitignored build directory. The script backgrounds this process.
3. The script starts `tsc` in watch mode on the client source and the view output directory, which is separate. This is also backgrounded.
4. The script starts rollup in watch mode on the output from tsc, which will now include the compiled templates.

Here's the nodemon watcher:

```js
const nodemon = require('nodemon');
const path = require('path');
const { build } = require('ractive/lib/component');
const fs = require('fs-extra');

const dir = 'views';
const outdir = 'src/node_modules';
const prefix = `${path.resolve('.')}/`;
let first = true;

const deps = {};

function mkReadFile(file) {
  const base = path.dirname(file);
  return function readFile(names) {
    if (Array.isArray(names)) names = path.join.apply(path, names);
    const name = path.isAbsolute(names) ? names : path.join(base, names);
    if (!~deps[file].indexOf(name)) deps[file].push(name);
    return fs.readFile(name, { encoding: 'utf8' });
  };
}

async function rebuild(f, nodep, rebuilt) {
  const file = f.replace(prefix, '');
  if (rebuilt && ~rebuilt.indexOf(file)) return;
  const target = path.join(outdir, file.replace('.ractive.html', '.ts'));
  console.log(`views: building ${file}`);
  const content = await fs.readFile(file, { encoding: 'utf8' });
  deps[file] = [];
  const output = await build(content, { autoExport: true, escapeUnicode: true }, mkReadFile(file));
  await fs.mkdirp(path.dirname(target));
  fs.writeFile(target, output, { encoding: 'utf8' });
  if (!nodep) {
    rebuilt = rebuilt || [];
    rebuilt.push(file);
    for (const k in deps) if (~deps[k].indexOf(file)) await rebuild(k, false, rebuilt);
  }
}

async function lsr(p) {
  const dir = (await fs.readdir(p)).map(f => path.join(p, f));
  let res = dir.slice();
  for (const f of dir) if ((await fs.stat(f)).isDirectory()) res = res.concat(await lsr(f));
  return res;
}

let cache = [];
let tm;

function flush() {
  const files = cache;
  cache = [];
  tm = null;
  const rebuilt = [];
  files.forEach(f => rebuild(f, false, rebuilt));
}

// change the watch array if your views aren't in the views dir
nodemon({ watch: ['views'], verbose: true, ext: 'html', exec: 'false' }).
  on('start', async () => {
    if (first) {
      first = false;
      const files = (await lsr(dir)).filter(f => /\.ractive\.html$/.test(f));
      for (let i = 0; i < files.length; i++) await rebuild(files[i], true);
      console.log('views initial build complete');
    }
  }).on('restart', files => {
    let i = files.length;
    while (i--) if (!~cache.indexOf(files[i])) cache.push(files[i]);
    if (tm) clearTimeout(tm);
    tm = setTimeout(flush, 500);
  }).on('stop', () => {
    console.log('done watching views');
    process.exit(0);
  });

process.on('SIGINT', () => {
  nodemon.emit('stop');
});
```

**Comment by DillonSadofsky** [comment]
That's interesting.  In our project, we've been able to use rollup plugins to chain everything, and that has seemed really nice.  I even wrote a couple custom rollup plugins to do various generated code translations 'before' typescript (which we use the rollup typescript plugin to do).

I suppose we could un-rollup the ractive template build step, but its so close to being able to watch our source code and generate the bundled output, cradle to grave.  The only thing that seems to not be working is that the ractive template rollup plugin writes output files successfully on the input files changing, but even though it goes 'before' typescript in the plugin order, the updated files are not present by the time typescript plugin runs.

My guess is that you generate and write the output view files to disk asynchronously and therefore they are not available to the next plugin by the time its invoked in the chain.  Would it be possible to change that?  Its nice that there is this plugin to rollup that can piggy back on the file watching and generate output, but if its not guaranteed as input to the next step in the rollup chain, it doesn't work quite right.

I think when I was reviewing the typescript rollup plugin some plugins will generate 'virtual' output that can be passed on to the next plugin (essentially an in-memory file), so at least if that could be returned to the rollup handler when it goes to 'look' for the generated file that would fix it, even if the files won't be flushed to disk until later.

**Reply by evs-chris** [reply]
I think the reason I didn't go with the rollup plugin for this project is I couldn't get incremental support working at the time with the build complexity and I have ton of template to run through. Also, incremental tsc via rollup was not great.

Are you using the `rollup-plugin-ractive-bin@2.0.0-alpha-3`? I have a few smaller projects running that, and for the most part, they work fine with templates being consumed by a later tsc step. Looking through [the plugin code](https://github.com/ractivejs/rollup-plugin-ractive-bin/blob/master/index.js), I don't see anywhere it's not using a promise to return after processing the templates or after writing a file out.

[This project](https://github.com/evs-chris/raport/blob/master/rollup.config.js) has a slightly weird build, but it uses the rollup plugin fairly successfully. I've noticed that occasionally I'll need to re-save a template for a change to show up. I think there may be some edge case rollup bugs involved though, because I had to add a timeout check on a package build because rollup will hang on the last step about half the time.

**Reply by DillonSadofsky** [reply]
Yeah, we're using 2.0.0-alpha-3. Yeah, reviewing your plugin, it seems like its probably not a race condition, so IDK why I'm so consistently getting the correct output files written to disk, but the 'old' versions consumed by the next plugin?  Maybe it has to do with what rollup timing 'handler' the event is happening on?

I have a simple rollup plugin that generates files to be consumed later, but I did it the 'dumb' way and just do it all in `buildStart` of the plugin vs as you go.  The only other thing I can think of is during execution I both hit ctx.addWatchFile for the containing directories via glob (to catch added and removed files) and also the generated output files.  I can't remember why I automatically added the generated output files to the watch list or if that has any effect on the current rollup execution?  Maybe if you called addWatchFile on the files that get generated from fs.writeFile?

IDK, rollup can be tough to debug.

**Reply by DillonSadofsky** [reply]
I'm still trying some stuff, but I added some log lines to the plugin js filtered to just one of my view files to compare the first launch (which seems to work in order) with the incremental build (which seems to not always work).  In the following, I'm modifying a string in a template file `change-password.ractive.html` then reviewing the same string in the `change-password.ractive.html.ts` and my output `build.js` at various breakpoints.  As you can see, typescript seems to always run synchronously after your resolveFile, so far so good:

```txt
Debugger attached.
rollup v2.53.2
bundles ./client/app.js → static\js\build.js...
resolveId(client/components/change-password.ractive.html) called in ractive-rollup
resolveFile(client/components/change-password.ractive.html) called in ractive-rollup
load(C:\Work\projects\presage-web\client\.views\client\components\change-password.ractive.html.ts) called in ractive-rollup
load(C:\Work\projects\presage-web\client\.views\client\components\change-password.ractive.html.ts) called in typescript-rollup
created static\js\build.js in 15.9s

[2022-03-01 14:49:15] waiting for changes...
bundles ./client/app.js → static\js\build.js...
load(C:\Work\projects\presage-web\client\.views\client\components\change-password.ractive.html.ts) called in ractive-rollup
resolveFile(client/components/change-password.ractive.html) called in ractive-rollup
load(C:\Work\projects\presage-web\client\.views\client\components\change-password.ractive.html.ts) called in typescript-rollup
created static\js\build.js in 1.5s

[2022-03-01 14:49:21] waiting for changes...
```

So why does this work on first run, but not on incremental build?  I set some breakpoints in the typescript rollup plugin and stepped in when it hit `load` for my file.  Apparently it keeps a cache of 'emitted files' (the .js and .js.map files) based on input .ts files, and it returns that.  I'm still trying to figure out why typescript ends up compiling the file in question, but not until after it retrieves the cached version during module resolution.  Anyway, IDK if this comes down to `@rollup/plugin-typescript` being overly aggressive with its cache or if there is some operation that needs to be called on the 'modified' input files that would notify ts that the file has changed.  Maybe you should be calling `watchChanged` manually when you rebuild the view .ts files? Maybe that would dirty ts' cache.  I'll try some stuff.
IDK if this is the right place to debug this.  I can take it elsewhere if you'd prefer.

**Reply by DillonSadofsky** [reply]
I'm starting to think this is a problem with the typescript rollup package, as I set a breakpoint in the code where it actually builds, and that seems to trigger between when we write the files out and when the final rolled up file is emitted.  Stepping through I saw this code in typescript.js which seems like a potential culprit:
```js
        // Upon detecting a file change, wait for 250ms and then perform a recompilation. This gives batch
        // operations (such as saving all modified files in an editor) a chance to complete before we kick
        // off a new compilation.
        function scheduleProgramUpdate() {
            if (!host.setTimeout || !host.clearTimeout) {
                return;
            }
            if (timerToUpdateProgram) {
                host.clearTimeout(timerToUpdateProgram);
            }
            writeLog("Scheduling update");
            timerToUpdateProgram = host.setTimeout(updateProgramWithWatchStatus, 250);
        }
```

This is called by a file watcher, so I think the narrative is that input view file modified->kicks off rollup->runs ractive view builder, which writes .ts file->internal tsc watcher, which decides not to compile yet, but schedule doing that for 250ms in the future->out of date cache of view .ts file is used->depending on race condition, internal typescript cache is updated

So there seems to be a quarter second race condition in the `tsc -w` being used internally by the `@rollup/typescript`.  Maybe some projects don't have this issue  because they're big enough that 'enough time passes' between .ts files being written and before typescript is invoked.  I guess a sleep (250) before continuing but after generating ractive templates could maybe fix it, but man that is ugly.

**Reply by DillonSadofsky** [reply]
Yuck.  I think I am right.  I added a sleep to rollup-plugin-ractive-bin\index.js right after the file write and it worked, though it slowed down the build (obviously):
![image](https://user-images.githubusercontent.com/2467337/156264331-55409961-3d8e-4a92-a7e4-e257d79db2ec.png)

I'm not sure of a good way to work around this race condition between files being output and read in by typescript.  I think I've been able to avoid it because my rollup plugins are stupid and just do everything in `buildStart` which must be separated enough timing-wise.  I guess I don't know if there is a good way around this.  Even the 'rollup twice' solution might not work if it happens *fast enough* because of the cache delay.  I know this is now only tertiarily your project's problem, but curious if you have any suggestions.

**Reply by evs-chris** [reply]
Ahhh, yes, now I remember. I couldn't get the typescript or typescript2 plugins to compile stuff and check types still and not have to be goaded to compile two or three times per save. So I switched my setup around to use a separate tsc process, which removes a bit of the faffery with rollup and caching. It's not ideal, but it's a lot smoother.

We could add a plugin option to chill for `x`ms after running. That's effectively what you've done, but I imagine that the ideal number may change from project to project. Since the plugin ostensibly supports ts by allowing you to change the extension, it would make sense to try to work a little better with the typescript plugins.

**Reply by DillonSadofsky** [reply]
Well, I wrote this up for them, since I guess its really their bug: https://github.com/rollup/plugins/issues/1122
IDK if they'll just close it since I didn't submit a minimal repro (due to race conditions being inherently difficult to repro).  I think the issue with the delays is that your rollup plugin will build .ts files 'one at a time' as part of `load` being called, and each one will hit the typescript's watch code, potentially having a 250ms delay, then the typescript rollup plugin's `load` is called, interleaved.  So since we don't do 'all ractive' then 'all typescript' we can't do just a single 250ms delay, to be safe, we have to wait 250ms after every view file is generated by the view generation code, which could really stack up on a large project.

I'm starting to think the safest solution would be to have your module generate .js code instead of .ts code so that typescript will not look for it, and it will be on disk for rollup to find and bundle later.  Though then I can't 'import' the output of the ractive transpile.  IDK, I'd have to experiment with that, and it feels like just working around the issue instead of solving it.

For now, I think the 'build twice' thing will work since rollup is slow and therefore always going to be slower than the 250ms delay, but if they ever sped it up, build twice might not work right.

**Reply by evs-chris** [reply]
Do you think trying to dump all of the output to disk at once would be a viable workaround? It would probably be possibly to have a global debounce that gates everything that happens within a certain window such that they return as a sort of batch.

**Reply by DillonSadofsky** [reply]
I do think that would technically resolve the issue, but it means making a 'dumber' rollup plugin.  The rollup plugins I made are super dumb and just do their work as part of the `buildStart` callback in rollup, but the downside is that you have to explore the source tree with globs to decide what files to 'do' vs handling them just in time as part of `resolveId` or `load`.  The downside of this 'dumb' way is you have to process all `/**/*.ractive.html` or whatever.  The nice thing about the way you're doing it is that rollup explores the source graph and only asks you to process files as they are 'hit'.

So I think that would work, but it feels kinda sucky to have to make that transition.
